File: py/0116_Populating_Next_Right_Pointers_in_Each_Node.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


class Node:
    def __init__(self, val: int = 0, left: "Node" = None, right: "Node" = None, next: "Node" = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


class Solution:
    def connect(self, root: "Node") -> "Node":
        if not root:
            return root

        root_ptr = root

        while root_ptr.left:
            ptr = root_ptr

            while ptr:
                ptr.left.next = ptr.right

                if ptr.next:
                    ptr.right.next = ptr.next.left

                ptr = ptr.next

            root_ptr = root_ptr.left

        return root


if __name__ == "__main__":
    # solution = Solution()

    sys.exit(0)

File: cpp/0116_Populating_Next_Right_Pointers_in_Each_Node.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution {
public:
    Node* connect(Node* root) {
        if (!root) return nullptr;
        queue<Node *> q;
        q.push(root);

        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                Node *node = q.front(); q.pop();

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);

                node->next = (i == size - 1) ? nullptr : q.front();
            }
        }

        return root;
    }
};

int main(int argc, char *argv[]) {

    return 0;
}
